src/stakeinput.cpp

```diff
@@ -9,67 +9,60 @@
 #include "zpiv/deterministicmint.h"
 #include "wallet/wallet.h"
 
-bool CPivStake::InitFromTxIn(const CTxIn& txin)
+CPivStake* CPivStake::NewPivStake(const CTxIn& txin)
 {
-    if (txin.IsZerocoinSpend())
-        return error("%s: unable to initialize CPivStake from zerocoin spend", __func__);
+    if (txin.IsZerocoinSpend()) {
+        error("%s: unable to initialize CPivStake from zerocoin spend", __func__);
+        return nullptr;
+    }
 
     // Find the previous transaction in database
     uint256 hashBlock;
     CTransaction txPrev;
-    if (!GetTransaction(txin.prevout.hash, txPrev, hashBlock, true))
-        return error("%s : INFO: read txPrev failed, tx id prev: %s", __func__, txin.prevout.hash.GetHex());
-    SetPrevout(txPrev.vout[txin.prevout.n], txin.prevout);
+    if (!GetTransaction(txin.prevout.hash, txPrev, hashBlock, true)) {
+        error("%s : INFO: read txPrev failed, tx id prev: %s", __func__, txin.prevout.hash.GetHex());
+        return nullptr;
+    }
 
+    const CBlockIndex* pindexFrom = nullptr;
     // Find the index of the block of the previous transaction
     if (mapBlockIndex.count(hashBlock)) {
         CBlockIndex* pindex = mapBlockIndex.at(hashBlock);
         if (chainActive.Contains(pindex)) pindexFrom = pindex;
     }
     // Check that the input is in the active chain
-    if (!pindexFrom)
-        return error("%s : Failed to find the block index for stake origin", __func__);
-
-    // All good
-    return true;
-}
-
-bool CPivStake::SetPrevout(const CTxOut& out, const COutPoint& outpointFrom)
-{
-    this->opOutputFrom = out;
-    this->opOutpointFrom = outpointFrom;
-    return true;
-}
+    if (!pindexFrom) {
+        error("%s : Failed to find the block index for stake origin", __func__);
+        return nullptr;
+    }
 
-void CPivStake::SetIndexFrom(const CBlockIndex* pindex)
-{
-    this->pindexFrom = pindex;
+    const CTxOut opOutputFrom = txPrev.vout[txin.prevout.n];
+    const COutPoint opOutpointFrom = txin.prevout;
+    return new CPivStake(opOutputFrom, opOutpointFrom, pindexFrom);
 }
 
 bool CPivStake::GetTxOutFrom(CTxOut& out) const
 {
-    if (!opOutputFrom || !opOutpointFrom)
-        return false;
-    out = *opOutputFrom;
+    out = opOutputFrom;
     return true;
 }
```

Comment:
Without optionals, you can make this a simple getter now
```cpp
CTxOut GetTxOutFrom() const { return opOutputFrom; }
```

Reply:
I thought on it as well but it's not as simple as changing it here. The method it's part of the base class. zpiv stake object needs to implement it as well and.. zpiv stake doesn't have an out from so.. it will need to return an empty `CTxOut`.

In other words, it will require a little bit larger refactor to make it properly. Remove the method from the base class, cast the subclass and call this `GetTxOutFrom`.

Reply:
No need to cast. You can simply return an empty CTxOut on zpiv stakes, it's not used there anyway.

Reply:
Yep, i said that at the end of my first paragraph :p.

I'm not really in favor of returning an empty object because some code is not currently using it. 
Prefer to leave it as is now and directly remove the method and make the flow cleaner in a subsequent PR.